From 87057fe9dcb1d9e261cb2344861aa3cffb9fa76b Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Wed, 20 Nov 2024 01:13:04 +0100
Subject: [PATCH] [DogeCash] Add `CalcExpectedMerkleTreeIndex` (#48)

This function allows tweaking the indices at which the hash of our block occurs in the merkle tree in the merge-mined block.

The design of this seems a bit obscure, and the merge mine nonce is 0 for most blocks on the network, but it's a consensus rule so we'll have to follow it exactly.

(cherry picked from commit bd97535bca2910534f5a082b618d25b0863328e3)

diff --git a/src/primitives/auxpow.cpp b/src/primitives/auxpow.cpp
index e9dab55647dd9baff843cf7ce7fe2ecec1dae0d4..9bf899ba85523a9df700131209accfb4b374e454 100644
--- a/src/primitives/auxpow.cpp
+++ b/src/primitives/auxpow.cpp
@@ -121,3 +121,28 @@ ParsedAuxPowCoinbase::Parse(const CScript &scriptCoinbase, uint256 hashRoot) {
         .nMergeMineNonce = nMergeMineNonce,
     }};
 }
+
+uint32_t CalcExpectedMerkleTreeIndex(uint32_t nNonce, uint32_t nChainId,
+                                     uint32_t merkleHeight) {
+    // Choose a pseudo-random slot in the chain merkle tree but have it be fixed
+    // for a size/nonce/chain combination.
+    // This prevents the same work from being used twice for the same chain
+    // while reducing the chance that two chains clash for the same slot.
+
+    // This computation can overflow the uint32 used. This is not an issue,
+    // though, since we take the mod against a power-of-two in the end anyway.
+    // This also ensures that the computation is, actually, consistent even
+    // if done in 64 bits as it was in the past on some systems.
+    // Note that h is always <= 30 (enforced by the maximum allowed chain merkle
+    // branch length), so that 32 bits are enough for the computation.
+
+    const uint32_t TWIST_FACTOR = 1103515245;
+    const uint32_t TWIST_OFFSET = 12345;
+
+    uint32_t rand = nNonce;
+    rand = rand * TWIST_FACTOR + TWIST_OFFSET;
+    rand += nChainId;
+    rand = rand * TWIST_FACTOR + TWIST_OFFSET;
+
+    return rand % (1 << merkleHeight);
+}
diff --git a/src/primitives/auxpow.h b/src/primitives/auxpow.h
index 98ab4795df8f9933578d31d838c4dfa08b2cf4d9..d880939557791339e63d4f5f8a26fc6109cbaf4c 100644
--- a/src/primitives/auxpow.h
+++ b/src/primitives/auxpow.h
@@ -85,6 +85,13 @@ uint256 ComputeMerkleRootForBranch(uint256 hash,
                                    const std::vector<uint256> &vMerkleBranch,
                                    uint32_t nIndex);
 
+/*
+ * Choose a pseudo-random slot in the chain merkle tree but have it be fixed for
+ * a size/nonce/chain combination.
+ */
+uint32_t CalcExpectedMerkleTreeIndex(uint32_t nNonce, uint32_t nChainId,
+                                     uint32_t merkleHeight);
+
 /** Parsed data from a AuxPow coinbase */
 class ParsedAuxPowCoinbase {
 public:
diff --git a/src/test/dogecoin_auxpow_check_tests.cpp b/src/test/dogecoin_auxpow_check_tests.cpp
index c7766b2cda88e1dc6205929fac7bd6d1b37c9324..132463b800e04a2315e3522131ad09147222bce2 100644
--- a/src/test/dogecoin_auxpow_check_tests.cpp
+++ b/src/test/dogecoin_auxpow_check_tests.cpp
@@ -94,6 +94,11 @@ BOOST_AUTO_TEST_CASE(auxpow_block700000_test) {
 
     BOOST_CHECK_EQUAL(parsed.nTreeSize, 64);
     BOOST_CHECK_EQUAL(parsed.nMergeMineNonce, 0);
+    BOOST_CHECK_EQUAL(
+        CalcExpectedMerkleTreeIndex(parsed.nMergeMineNonce,
+                                    VersionChainId(header.nVersion),
+                                    auxpow.vChainMerkleBranch.size()),
+        56);
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_block800000_test) {
@@ -110,6 +115,11 @@ BOOST_AUTO_TEST_CASE(auxpow_block800000_test) {
 
     BOOST_CHECK_EQUAL(parsed.nTreeSize, 64);
     BOOST_CHECK_EQUAL(parsed.nMergeMineNonce, 2677055472);
+    BOOST_CHECK_EQUAL(
+        CalcExpectedMerkleTreeIndex(parsed.nMergeMineNonce,
+                                    VersionChainId(header.nVersion),
+                                    auxpow.vChainMerkleBranch.size()),
+        40);
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_block3000000_test) {
@@ -126,6 +136,11 @@ BOOST_AUTO_TEST_CASE(auxpow_block3000000_test) {
 
     BOOST_CHECK_EQUAL(parsed.nTreeSize, 16);
     BOOST_CHECK_EQUAL(parsed.nMergeMineNonce, 0);
+    BOOST_CHECK_EQUAL(
+        CalcExpectedMerkleTreeIndex(parsed.nMergeMineNonce,
+                                    VersionChainId(header.nVersion),
+                                    auxpow.vChainMerkleBranch.size()),
+        8);
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_parse_coinbase_test) {
@@ -309,4 +324,18 @@ BOOST_AUTO_TEST_CASE(auxpow_parse_coinbase_test) {
     }
 }
 
+BOOST_AUTO_TEST_CASE(CalcExpectedMerkleTreeIndex_test) {
+    // Block 564415
+    BOOST_CHECK_EQUAL(CalcExpectedMerkleTreeIndex(0, AUXPOW_CHAIN_ID, 5), 24);
+    // Block 750100
+    BOOST_CHECK_EQUAL(
+        CalcExpectedMerkleTreeIndex(0x77654e2f, AUXPOW_CHAIN_ID, 6), 63);
+    // Block 805660
+    BOOST_CHECK_EQUAL(
+        CalcExpectedMerkleTreeIndex(0x9f909ff0, AUXPOW_CHAIN_ID, 6), 40);
+    // Block 845783
+    BOOST_CHECK_EQUAL(CalcExpectedMerkleTreeIndex(0, AUXPOW_CHAIN_ID, 11),
+                      1080);
+}
+
 BOOST_AUTO_TEST_SUITE_END()
